numpy/core/tests/test_umath.py

```diff
@@ -650,6 +649,34 @@ def test_exp_values(self):
             yf = np.array(y, dtype=dt)*log2_
             assert_almost_equal(np.exp(yf), xf)
 
+class TestSpecialFloats(object):
+    def test_exp_values(self):
+        x = [np.nan,  np.nan, np.inf, 0.]
+        y = [np.nan, -np.nan, np.inf, -np.inf]
+        for dt in ['f', 'd', 'g']:
+            xf = np.array(x, dtype=dt)
+            yf = np.array(y, dtype=dt)
+            assert_equal(np.exp(yf), xf)
```

Comment:
This is prone to failure on my mac locally:

```
__________________________________________________________________________________ TestSpecialFloats.test_exp_values ___________________________________________________________________________________
[gw0] darwin -- Python 3.6.5 /Users/treddy/miniconda3/envs/numpy_dev_py36/bin/python

self = <numpy.core.tests.test_umath.TestSpecialFloats object at 0x11ef99ba8>

    def test_exp_values(self):
        x = [np.nan,  np.nan, np.inf, 0.]
        y = [np.nan, -np.nan, np.inf, -np.inf]
        for dt in ['f', 'd', 'g']:
            xf = np.array(x, dtype=dt)
            yf = np.array(y, dtype=dt)
>           assert_equal(np.exp(yf), xf)
E           RuntimeWarning: invalid value encountered in exp

dt         = 'f'
self       = <numpy.core.tests.test_umath.TestSpecialFloats object at 0x11ef99ba8>
x          = [nan, nan, inf, 0.0]
xf         = array([nan, nan, inf,  0.], dtype=float32)
y          = [nan, nan, inf, -inf]
yf         = array([ nan,  nan,  inf, -inf], dtype=float32)

numpy/core/tests/test_umath.py:659: RuntimeWarning
1 failed, 8142 passed, 63 skipped, 13 xfailed, 3 xpassed in 131.72 seconds
```

Reply:
Is your macos setup significantly different to the azure CI one (clang/OS version)?

Reply:
`gcc --version`
```
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin18.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```
That means `Xcode 10.2.1` according to [these tables](https://en.wikipedia.org/wiki/Xcode)

`sw_vers -productVersion`
`10.14.4`

By contrast, Azure is `10.13` series OS and perhaps slightly different Xcode 10 minor version? 

There is apparently now an option to bump Azure Mac CI to `macOS-10.14`--not sure we want to deal with that right now, but easy experiment I suppose. It may be the Xcode version--there are docs about switching these things on Azure, depending on availability.

Reply:
do we know a fix for this? I am unable to reproduce this on my ubuntu and gcc/clang. one possible solution is: since that portion of test `test_exp_values `only cares about verifying the output of exp, I can may be ignore floating point errors by updating the test with `np.errstate(all='ignore')`. Would that be acceptable?

Reply:
I would suggest the code explicitly mask the warnings after the call to the new functions via `npy_clear_floatstatus_barrier((char*)dimensions)`, see the other uses in `loops.c.src`. @tylerjereddy could you test that on your machine? If it helps, we could put it into a #ifdef WHATEVER_CLANG_WE_NEED block

Reply:
But that would lead to these portions of the tests failing though, right?  
```
assert_raises(FloatingPointError, np.exp, np.float32(100.)) #overflow 
assert_raises(FloatingPointError, np.exp, np.float32(1E19)) #overflow 
assert_raises(FloatingPointError, np.log, np.float32(-np.inf)) #invalid 
assert_raises(FloatingPointError, np.log, np.float32(-1.0)) #invalid 
assert_raises(FloatingPointError, np.log, np.float32(0.)) #divide by zero 
```

Reply:
If you provide a `git diff` on the patch you want me to try I can do that; probably even better to try reproducing in Azure MacOS on a fork or something, but in my experience that can take a few hours at least to iterate / get right. 

I tried a quick bump of the Azure MacOS version alone to `10.14`, which is documented to use the same Xcode version I describe, but it didn't reproduce immediately. Probably needs playing around with some flags, clang version, or something else I haven't thought of yet..